src/main/java/de/tum/in/www1/artemis/service/ExerciseService.java

```diff
@@ -765,4 +776,113 @@ public void validateScoreSettings(Exercise exercise) {
             throw new BadRequestAlertException("The provided bonus points are not allowed", "Exercise", "bonusPointsInvalid");
         }
     }
+
+    /**
+     * Checks the exercise grading instructions if any of them is associated with the feedback
+     * then, sets the corresponding exercise field
+     *
+     * @param gradingCriteria grading criteria list of exercise
+     * @param exercise exercise to update
+     *
+     */
+    public void checkExerciseIfGradingInstructionFeedbackUsed(List<GradingCriterion> gradingCriteria, Exercise exercise) {
+        List<Feedback> feedbackList = feedbackRepository.findFeedbackByStructuredGradingInstructionId(gradingCriteria);
+
+        if (!feedbackList.isEmpty()) {
+            exercise.setGradingInstructionFeedbackUsed(true);
+        }
+    }
+
+    /**
+     * Re-evaluates the exercise before saving
+     * 1. The feedback associated with the exercise grading instruction needs to be updated
+     * 2. After updating feedbacks, result needs to be re-calculated
+     *
+     * @param exercise exercise to re-evaluate
+     * @param deleteFeedbacks  about checking if the feedbacks should be deleted when the associated grading instructions are deleted
+     *
+     */
+    public void reEvaluateExercise(Exercise exercise, Boolean deleteFeedbacks) {
+
+        List<GradingCriterion> gradingCriteria = exercise.getGradingCriteria();
+        // retrieve the feedbacks associated with the grading instructions
+        List<Feedback> feedbackList = feedbackRepository.findFeedbackByStructuredGradingInstructionId(gradingCriteria);
+
+        // collect each sub-grading instructions into the list
```

Comment:
ah, I see the confusion. Grading criterion is like wrapper for grading instructions. Since, grading instructions are connected with exercise via Grading Criterion. I did not want to loop for each grading criterion and grading instruction respectively. So, i wanted to collect all of the grading instructions that belong to exercise in one list. I can just leave "sub", i think....